include/exadg/fluid_structure_interaction_precice/coupling_interface.h

```diff
@@ -0,0 +1,224 @@
+   */
+  virtual void
+  process_coupling_mesh();
+
+  /**
+   * @brief write_data Write the data associated to the defined vertice
+   *        to preCICE
+   *
+   * @param data_vector Vector holding the global solution to be passed to
+   *        preCICE.
+   */
+  virtual void
+  write_data(const LinearAlgebra::distributed::Vector<double> & data_vector,
+             const std::string &                                data_name) = 0;
+
+
+  virtual std::vector<Tensor<1, dim>>
+  read_block_data(const std::string & data_name) const;
+
+  /**
+   * @brief Queries data IDs from preCICE for the given read data name
+   * @param read_data_name
+   */
+  void
+  add_read_data(const std::string & read_data_name);
+
+  /**
+   * @brief Queries data IDs from preCICE for the given write data name
+   * @param write_data_name
+   */
+  void
+  add_write_data(const std::string & write_data_name, const std::string & write_data_specification);
+
+protected:
+  /**
+   * @brief Print information of the current setup
+   *
+   * @param[in] reader Boolean in order to decide if we want read or write
+   *            data information
+   * @param[in] local_size The number of element the local process works on
+   */
+  void
+  print_info(const bool reader, const unsigned int local_size) const;
+
+  /// The MatrixFree object (preCICE can only handle double precision)
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> mf_data;
+
+  /// public precice solverinterface
+  std::shared_ptr<precice::SolverInterface> precice;
+
+  /// Configuration parameters
+  const std::string mesh_name;
+  int               mesh_id = -1;
+  // Map between data ID (preCICE) and the data name
+  std::map<std::string, int> read_data_map;
+  std::map<std::string, int> write_data_map;
+
+  const types::boundary_id dealii_boundary_interface_id;
+
+  WriteDataType write_data_type = WriteDataType::undefined;
+
+  virtual std::string
+  get_interface_type() const = 0;
+};
+
+
+
+template<int dim, int data_dim, typename VectorizedArrayType>
+CouplingInterface<dim, data_dim, VectorizedArrayType>::CouplingInterface(
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> data,
+  std::shared_ptr<precice::SolverInterface>                           precice,
+  std::string                                                         mesh_name,
+  const types::boundary_id                                            interface_id)
+  : mf_data(data),
+    precice(precice),
+    mesh_name(mesh_name),
+    dealii_boundary_interface_id(interface_id)
+{
+  Assert(data.get() != nullptr, ExcNotInitialized());
+  Assert(precice.get() != nullptr, ExcNotInitialized());
+
+  // Ask preCICE already in the constructor for the IDs
+  mesh_id = precice->getMeshID(mesh_name);
+}
+
+
+template<int dim, int data_dim, typename VectorizedArrayType>
+void
+CouplingInterface<dim, data_dim, VectorizedArrayType>::add_read_data(
+  const std::string & read_data_name_)
```

Comment:
before merging, we should use a consistent naming with or without `_` at the end.